docs/proposals/202003_thanos_rules_federation.md

```diff
@@ -0,0 +1,81 @@
+
+Since we work with metrics, evaluating recording rules and alerts is a very important part of our system. Some of our "StoreAPIs" like Prometheus and Thanos Ruler are designed for that. However, we currently don’t have a way to present those resources in a federated way e.g Alerts and Recording Rules. This document explores the potential way of solving this in the Prometheus Ecosystem.
+
+## Motivation
+
+Deployment scenarios with various levels of federated setups are not unusual. Concrete deployments i.e. leverage Thanos Ruler to spawn multiple query sources (i.e. via Thanos Querier) for rule and alert evaluation. This topology sacrifices availability. At the same time in the same deployment topology a dedicated Prometheus instance is deployed for certain critical alerting rules implying high availability. Here, we are missing a federation middleware exposing a consolidated view of evaluated rules and alerts in the underlying Prometheus and ThanosRuler instances.
+
+Pitfalls of current solutions:
+* It’s tedious to manually visit each leaf and we are lazy.
+* Visiting leaf instances (be it ThanosRuler or dedicated Prometheus endpoints) have to be implemented manually.
+
+## Goals
+
+* Present other Prometheus/Thanos Resources in a global view.
+    * This also means an up-to-date view (e.g statuses)
+* It can be tied to Thanos.
+* Simple to run if you already run Thanos.
+    * Use consistent protocols (if we suddenly switch to pure HTTP from gRPC it will break user’s proxies for auth, routing, monitoring, rate limiting, etc)
+    * Potentially use the same connections, discovered targets by Querier.
+
+## Verification
+
+* Unit tests which would fire up a dummy RulesAPI and check different scenarios.
+* Ad-hoc testing.
+
+## Proposal
+
+We propose to add following proto service and propagate all using existing components that has rules, so:
+* Sidecar
+* Querier (for federation)
+* Ruler
+
+The newly introduced `Rules` service is designed to reflect [Prometheus' Rules API](https://github.com/prometheus/prometheus/blob/bc703b64568ebfaecf27b9b70be737ad318e217a/web/api/v1/api.go) allowing to retrieve recording and alertings rules.
+
+```proto
+service Rules {
+  /// Rules has info for all rules.
+  rpc Rules(RulesRequest) returns (stream RulesResponse);
+}
+```
+
+Sidecar proxies the Rules request to its local Prometheus instance and synthesizes the response. Similarly Thanos rule constructs the response based in its local state.
+
+Thanos query fans-out to all know Rules endpoints configured via `--rule` command line flag, merges and deduplicates the result. Generally the deduplication logic is less complex than with time series, specifically:
+
+* Deduplication happens first the rule group level. The identifier is the group name.
+* Then, per group name deduplication happens on the rule level, where:
+
+1. the rule type (recording rule vs. alerting rule)
+2. the rule name
+3. the rule label names
```

Comment:
Don't we need manual flag, like `--rule.replica-label=replica` be also part of identifier?

What I mean is scenario:

- Team1 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
- Team2 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff  which fires on their k8s namespace (different expr field, same labels)

In this case dedup would show a single rule?

I think this case might be quite common with monitoring mixins?

Reply:
Good catch, I believe this is the same edge case as https://github.com/thanos-io/thanos/pull/2263#discussion_r396521746.

I would suggest to add the expression as part of the deduplication key.

As for `--rule.replica-label=replica`: Good idea, but a counter-suggestion: Since rules api is a subset of store api, couldn't we reuse the same replica label, as the underlying metrics would have the same replica label? (just trying to prevent additional configuration complexity).

Reply:
Wonder if this is related: https://github.com/coreos/prometheus-operator/issues/2193

We need to decide if we want to reuse querier replica labels or add another flag. I think it's kind of complex to make it more difficult to configure replicas (another format). Why not just another flag? Let's discuss, maybe we can find some format that would work for us.

`--replica=alerts{replica}` etc? This is what you mean?

Reply:
Suggestion while hopefully retaining enough orthogonality:

1. We do introduce a suggested `--rule.replica-label=replica` cli argument.
2. If that argument is not given we fall back to `--query.replica-label`.

But I am also very fine with simply a completely independent cli arg.

Reply:
@povilasv question to clear out ambiguities: Do you think that by introducing `--rule.replica-label` cli arg does not necessitate to consider the expression be part of the deduplication logic?

Reply:
I think we still need expression, example alerts I have from kubernetes-mixin that have same exact labels names etc except for expression part:

```
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate1h) > (14.40 * 0.01000)
      and
      sum(apiserver_request:burnrate5m) > (14.40 * 0.01000)
    for: 2m
    labels:
      severity: critical
  - alert: KubeAPIErrorBudgetBurn
    annotations:
      message: The API server is burning too much error budget
    expr: |
      sum(apiserver_request:burnrate6h) > (6.00 * 0.01000)
      and
      sum(apiserver_request:burnrate30m) > (6.00 * 0.01000)
    for: 15m
    labels:
      severity: critical
```

Reply:
Indeed, very good example :+1: Adding the expression as part of the deduplication chain.

Reply:
In addition the `for` field should also be considered.

Reply:
Wow, I did not that is possible 0.o

Reply:
> Don't we need manual flag, like --rule.replica-label=replica be also part of identifier?
>
> What I mean is scenario:
>
> Team1 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
> Team2 runs Thanos Ruler and has an alert with name PodCrashLoopingBackOff which fires on their k8s namespace (different expr field, same labels)
> In this case dedup would show a single rule?
> 
> I think this case might be quite common with monitoring mixins?

I don't think we need separate replica labels. Let's not complicate things as,as you can see it can escalate quickly: https://github.com/thanos-io/thanos/pull/2263#discussion_r406109403

In the case you describes @povilasv Thanos Rule has to have unique labels as Prometheus external labels, so this is solved.